### ISSUE TITLE
DSpace 7.6.1: Bulk access control fails to iterate over each item in a collection, producing duplicate policies for some items 

### ISSUE THREAD

**Issue by ttoljamo** [DSpace/DSpace#9208]
**Describe the bug**
Bulk access control fails to iterate over each item in a collection and ends up producing duplicate policies for some items. 

As an administrator, I tried to add item-level read rights for a specific user group for all items in a collection. The collection was an owning collection containing 203983 items. After the process completed successfully, I noticed that not all items were discoverable by the target group. 

Upon examining the matter, I found that while the process log files contained the right amount policy adding statements (e.g. 203983 lines of `2023-11-21 14:23:32.830 INFO bulk-access-control - 13 @ Adding Item {29d339e7-830b-4cd5-bb70-a71bfe33dcbc} policy with access conditions: ...`), there were also several duplicate item uuids. For example, the item `29d84d8d-9230-4e51-afa7-b3f7bf1b490b` is mentioned in the log file 9 times, and true enough, the item now has 9 duplicate Read:Group policies. 

This wasn't an one-off incident: same behaviour happened when I tried using bulk access control on secondary collections with 40-60k items each. 

**To Reproduce**
1. Navigate to a collection containing a bunch of items 
2. Click 'Edit this collection' button
3. Select the 'Access control' leaf 
4. Select Item's Metadata
5. Select the Mode 'Add to existing ones' 
6. Select the Access condition type from a list
7. Click 'Execute'
8. Wait for the process to finish
9. Check that all items were given appropriate policies, e.g. items are readable by a specific user group, and examine the process log files. 

**Expected behavior**
Bulk access control should iterate over each item in the collection once and it should not produce duplicate policies. 

**Related work**

**Comment by tdonohue** [comment]
Pulling this over to the 7.6.x maintenance board.  It will need a volunteer to try to reproduce the issue or locate a possible cause.  Also may require more detail to help reproduce.  

@ttoljamo  If there's anything else you can share about the specific items that are duplicated, that may be helpful.  For instance if these Items are mapped to multiple collections, or have something else that is "unique" that could provide a very big clue as to what may be going on.  Otherwise, I'll need to find someone else able to reproduce the same issue.

**Comment by ttoljamo** [comment]
I had a look at a bunch of the duplicated items and could not spot any patterns that made those stand out from the rest. 

I don't have the exact numbers, but I should have emphasised that this affected a sizable portion of the items: perhaps 70-90k items did not get the read-policy from the bulk operation, and I suspect that more than 25k items now have duplicate policies.

All in all, the full set of 203k items is very uniform: 
- Each item belongs to the same owning collection. 
- Additionally, each item belongs to at least one other secondary collection, but may be mapped up to four others. 
- Each item contains exactly two bitstreams (JPEG images) in the original bundle, and two in the thumbnail bundle. 
- Before the bulk operation, each item had the same resource policies: read access for a specific group for items and bundles. Notably, neither items or bundles are read:anonymous, but the original bitstream images are (this is due to the image viewer bug). 
- Each item was imported using zipped SAF packages: first ~100k items were imported with 'handle' files so that they could keep their handles from another DSpace instance, but the rest got them assigned fresh during the import process. 
- Bitstreams are not imported within the SAF packages, but instead registered from two local bitstores. 
- Each item's metadata consists of fields from two schemas: dc and a custom schema. 

In summary, each item is very much alike all the others. If I had to venture to make a blind guess, I'd say that this could be some deep paging -related issue affecting the source of where the batch updater gets its identifiers. Maybe the ordering gets changed a bit during the process and the paging goes off.

**Comment by msahyoun** [comment]
from the ones where the read policy hasn't been set if you try Bulk Access Control on the item directly is the policy set correctly?

**Comment by msahyoun** [comment]
I wrote a component based of the source of BulkAccessManagement and am experiencing the same issue that the changes are not applied to all records. IMHO this is due to  issuing SOLR queries to fetch the items in an unsorted manner so that after applying changes to items, which updates the SOLR index  offset/start and limit are not fetching the next entries but the new search might skip or include already fetched ones.

**Comment by msahyoun** [comment]
Adding `discoverQuery.setSortField("search.resourceid", SORT_ORDER.asc);` in` private DiscoverQuery buildDiscoveryQuery` fixed the issue for me. 

IMHO this is because paged data has been used without sorting.

@tdonohue  WDYT?

**Comment by tdonohue** [comment]
@msahyoun : That does sound like a bug.  It definitely would be a bug to use pagination without sorting, and that *could* be the cause of objects being skipped in iteration (as the sort order may not be staying consistent as you use pagination).

I'd recommend creating a Pull Request with the fix you've applied.  Maybe @ttoljamo  (the original reporter) or others can help test it to verify it fixes the issue for them as well.  Thanks!

**Comment by mwoodiupui** [comment]
We probably should be using Solr cursors as well, in bulk operations.  https://solr.apache.org/guide/6_6/pagination-of-results.html

**Comment by msahyoun** [comment]
Being new to DSpace development I'm wondering why Bulk Access is using SOLR/DiscoverQuery to get the data where other tasks are using  e.g. ItemService. Could someone shed some light into this before I'm looking into creating a pull request. I also need to look at the documentation how you'd like these, coding standards ... so please bare with me for the relative small change.

Reg. Solr cursors - yes, that's an even better way to do it.
